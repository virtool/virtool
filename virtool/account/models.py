from datetime import datetime

from email_validator import EmailSyntaxError, validate_email
from pydantic import ConstrainedStr, validator

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.enums import AnalysisWorkflow
from virtool_core.models.group import GroupMinimal, Permissions
from virtool_core.models.user import User


def check_email(email: str | None) -> str | None:
    """Checks if the given email is valid."""
    if email is None:
        return None

    try:
        validate_email(email)
    except EmailSyntaxError:
        raise ValueError("The format of the email is invalid")

    return email


class AccountSettings(BaseModel):
    quick_analyze_workflow: AnalysisWorkflow
    show_ids: bool
    show_versions: bool
    skip_quick_analyze_dialog: bool


class ConstrainedEmail(ConstrainedStr):
    strip_whitespace = True


class Account(User):
    settings: AccountSettings
    email: ConstrainedEmail | None

    _email_validation = validator("email", allow_reuse=True)(check_email)


class APIKey(BaseModel):
    id: str
    created_at: datetime
    groups: list[GroupMinimal]
    name: str
    permissions: Permissions
