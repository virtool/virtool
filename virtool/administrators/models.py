from typing import List

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.roles import AdministratorRole
from virtool_core.models.user import UserNested


class AdministratorMinimal(UserNested):
    role: AdministratorRole


class Administrator(AdministratorMinimal):
    available_roles: List[dict]


class AdministratorSearch(BaseModel):
    items: List[AdministratorMinimal]
    available_roles: List[dict]
