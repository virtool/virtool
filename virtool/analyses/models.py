import enum
from sqlalchemy import Column, Integer, String, DateTime, Enum

from virtool.postgres import Base


class AnalysisFormat(str, enum.Enum):
    """
    Enumerated type for analysis file formats

    """

    sam = "sam"
    bam = "bam"
    fasta = "fasta"
    fastq = "fastq"
    csv = "csv"
    tsv = "tsv"
    json = "json"


class AnalysisFile(Base):
    """
    SQL model to store new analysis files

    """
    __tablename__ = "analysis_files"

    id = Column(Integer, primary_key=True)
    analysis = Column(String)
    description = Column(String)
    format = Column(Enum(AnalysisFormat))
    name = Column(String)
    name_on_disk = Column(String, unique=True)
    size = Column(Integer)
    uploaded_at = Column(DateTime)

    def __repr__(self):
        return f"<AnalysisFile(id={self.id}, analysis={self.analysis}, description={self.description}, " \
               f"format={self.format}, name={self.name}, name_on_disk={self.name_on_disk}, size={self.size}, " \
               f"uploaded_at={self.uploaded_at})>"
