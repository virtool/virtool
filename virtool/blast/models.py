from datetime import datetime

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.task import TaskNested


class NuvsBlast(BaseModel):
    id: int
    created_at: datetime
    updated_at: datetime
    last_checked_at: datetime
    error: str | None
    rid: str | None
    ready: bool = False
    result: dict | None
    task: TaskNested | None
