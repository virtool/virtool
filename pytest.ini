[pytest]
addopts = --aiohttp-loop uvloop --capture tee-sys
asyncio_mode = auto
