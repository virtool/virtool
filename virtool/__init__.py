"""Application module.

TODO: Remove import when spaces are completely removed.
"""

import virtool.spaces.sql  # noqa: F401
