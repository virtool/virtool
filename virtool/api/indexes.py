import virtool.db.history
import virtool.db.indexes
import virtool.db.jobs
import virtool.db.references
import virtool.db.utils
import virtool.history
import virtool.http.routes
import virtool.jobs.build_index
import virtool.utils
from virtool.api.utils import bad_request, compose_regex_query, conflict, insufficient_rights, json_response,\
    not_found, paginate

routes = virtool.http.routes.Routes()


@routes.get("/api/indexes")
async def find(req):
    """
    Return a list of indexes.

    """
    db = req.app["db"]

    ready = req.query.get("ready", False)

    if not ready:
        data = await virtool.db.indexes.find(db, req.query)
        return json_response(data)

    pipeline = [
        {
            "$match": {
                "ready": True
            }
        },
        {
            "$sort": {
                "version": -1
            }
        },
        {
            "$group": {
                "_id": "$reference.id",
                "index": {
                    "$first": "$_id"
                },
                "version": {
                    "$first": "$version"
                }
            }
        }
    ]

    ready_indexes = list()

    async for agg in db.indexes.aggregate(pipeline):
        reference_name = await virtool.db.utils.get_one_field(db.references, "name", agg["_id"])

        ready_indexes.append({
            "id": agg["index"],
            "version": agg["version"],
            "reference": {
                "id": agg["_id"],
                "name": reference_name
            }
        })

    return json_response(ready_indexes)


@routes.get("/api/indexes/{index_id}")
async def get(req):
    """
    Get the complete document for a given index.

    """
    db = req.app["db"]

    index_id = req.match_info["index_id"]

    document = await db.indexes.find_one(index_id)

    if not document:
        return not_found()

    document = virtool.utils.base_processor(document)

    document["contributors"] = await virtool.db.indexes.get_contributors(db, index_id)

    document["otus"] = await virtool.db.indexes.get_otus(db, index_id)

    document["change_count"] = sum(v["change_count"] for v in document["otus"])

    return json_response(document)


@routes.post("/api/refs/{ref_id}/indexes")
async def create(req):
    """
    Starts a job to rebuild the otus Bowtie2 index on disk. Does a check to make sure there are no unverified
    otus in the collection and updates otu history to show the version and id of the new index.

    """
    db = req.app["db"]

    ref_id = req.match_info["ref_id"]

    reference = await db.references.find_one(ref_id, ["groups", "users"])

    if reference is None:
        return not_found()

    if not await virtool.db.references.check_right(req, reference, "build"):
        return insufficient_rights()

    if await db.indexes.count({"reference.id": ref_id, "ready": False}):
        return conflict("Index build already in progress")

    if await db.otus.count({"reference.id": ref_id, "verified": False}):
        return bad_request("There are unverified OTUs")

    if not await db.history.count({"reference.id": ref_id, "index.id": "unbuilt"}):
        return bad_request("There are no unbuilt changes")

    index_id = await virtool.db.utils.get_new_id(db.indexes)

    index_version = await virtool.db.indexes.get_next_version(db, ref_id)

    job_id = await virtool.db.utils.get_new_id(db.jobs)

    manifest = await virtool.db.references.get_manifest(db, ref_id)

    user_id = req["client"].user_id

    document = {
        "_id": index_id,
        "version": index_version,
        "created_at": virtool.utils.timestamp(),
        "manifest": manifest,
        "ready": False,
        "has_files": True,
        "job": {
            "id": job_id
        },
        "reference": {
            "id": ref_id
        },
        "user": {
            "id": user_id
        }
    }

    await db.indexes.insert_one(document)

    await db.history.update_many({"index.id": "unbuilt", "reference.id": ref_id}, {
        "$set": {
            "index": {
                "id": index_id,
                "version": index_version
            }
        }
    })

    # A dict of task_args for the rebuild job.
    task_args = {
        "ref_id": ref_id,
        "user_id": user_id,
        "index_id": index_id,
        "index_version": index_version,
        "manifest": manifest
    }

    # Create job document.
    job = await virtool.db.jobs.create(
        db,
        req.app["settings"],
        "build_index",
        task_args,
        user_id,
        job_id=job_id
    )

    await req.app["jobs"].enqueue(job["_id"])

    headers = {
        "Location": "/api/indexes/" + index_id
    }

    return json_response(virtool.utils.base_processor(document), status=201, headers=headers)


@routes.get("/api/indexes/{index_id}/history")
async def find_history(req):
    """
    Find history changes for a specific index.

    """
    db = req.app["db"]

    index_id = req.match_info["index_id"]

    if not await db.indexes.count({"_id": index_id}):
        return not_found()

    term = req.query.get("term", None)

    db_query = {
        "index.id": index_id
    }

    if term:
        db_query.update(compose_regex_query(term, ["otu.name", "user.id"]))

    data = await paginate(
        db.history,
        db_query,
        req.query,
        sort=[("otu.name", 1), ("otu.version", -1)],
        projection=virtool.db.history.LIST_PROJECTION,
        reverse=True
    )

    return json_response(data)
