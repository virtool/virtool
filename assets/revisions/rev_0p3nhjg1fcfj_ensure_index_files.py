"""ensure index files

Revision ID: 0p3nhjg1fcfj
Date: 2024-05-22 20:47:09.866326

"""

from asyncio import to_thread
from pathlib import Path

import arrow
from motor.motor_asyncio import AsyncIOMotorDatabase
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from virtool_core.utils import file_stats

from virtool.api.custom_json import dump_bytes
from virtool.history.db import patch_to_version
from virtool.indexes.db import INDEX_FILE_NAMES
from virtool.indexes.models import IndexType, SQLIndexFile
from virtool.indexes.utils import join_index_path
from virtool.migration import MigrationContext
from virtool.types import Document
from virtool.utils import compress_json_with_gzip

# Revision identifiers.
name = "ensure index files"
created_at = arrow.get("2024-05-22 20:47:09.866326")
revision_id = "0p3nhjg1fcfj"

alembic_down_revision = None
virtool_down_revision = "ulapmx8i3vpg"


# Change this if an Alembic revision is required to run this migration.
required_alembic_revision = None


async def upgrade(ctx: MigrationContext):
    async for index in ctx.mongo.indexes.find({"ready": True}):
        index_id = index["_id"]

        index_path = join_index_path(
            ctx.data_path,
            index["reference"]["id"],
            index_id,
        )

        if not index_path.is_dir():
            continue

        try:
            await ensure_json(
                ctx.mongo,
                index_path,
                ctx.data_path,
                index["reference"]["id"],
                index["manifest"],
            )
        except IndexError:
            continue

        async with AsyncSession(ctx.pg) as session:
            first = (
                await session.execute(
                    select(SQLIndexFile).where(SQLIndexFile.index == index_id),
                )
            ).first()

            if first:
                continue

            session.add_all(
                [
                    SQLIndexFile(
                        name=path.name,
                        index=index_id,
                        type=get_index_file_type_from_name(path.name),
                        size=(await to_thread(file_stats, path))["size"],
                    )
                    for path in sorted(index_path.iterdir())
                    if path.name in INDEX_FILE_NAMES
                ],
            )

            await session.commit()


async def ensure_json(
    mongo: AsyncIOMotorDatabase,
    index_path: Path,
    data_path: Path,
    ref_id: str,
    manifest: dict,
):
    """Ensure that a there is a compressed JSON representation of the index found at
    `path`` exists.

    :param mongo: the application mongo client
    :param index_path: the path to the index directory
    :param data_path: the path to the data directory
    :param ref_id: the id of the parent reference
    :param manifest: the otu id-version manifest for the index
    """
    json_path = index_path / "reference.json.gz"

    if await to_thread(json_path.is_file):
        return

    reference = await mongo.references.find_one(
        ref_id,
        ["data_type", "organism", "targets"],
    )

    await to_thread(
        compress_json_with_gzip,
        dump_bytes(
            {
                "data_type": reference["data_type"],
                "organism": reference["organism"],
                "otus": await export_index(data_path, mongo, manifest),
                "targets": reference.get("targets"),
            },
        ),
        json_path,
    )


async def export_index(
    data_path: Path,
    mongo: "Mongo",
    manifest: dict[str, int],
) -> list[Document]:
    """Dump OTUs to a JSON-serializable data structure based on an index manifest.

    :param data_path: the application data path
    :param mongo: the application mongo client
    :param manifest: the index manifest
    :return: JSON-serializable OTU data
    """
    otu_list = []

    for otu_id, otu_version in manifest.items():
        _, joined, _ = await patch_to_version(data_path, mongo, otu_id, otu_version)
        otu_list.append(format_otu_for_export(joined))

    return otu_list


def format_otu_for_export(otu: Document) -> Document:
    """Prepare a raw, joined OTU for export.

    If the OTU has a remote ID use this as the `_id` for export. This makes the OTU
    compatible with the original remote reference.

    :param otu: a joined OTU document
    :return: the formatted, joined OTU document
    """
    try:
        otu_id = otu["remote"]["id"]
    except KeyError:
        otu_id = otu["_id"]

    isolates = [
        {
            "id": isolate["id"],
            "default": isolate["default"],
            "source_name": isolate["source_name"],
            "source_type": isolate["source_type"],
            "sequences": [
                format_sequence_for_export(sequence)
                for sequence in isolate["sequences"]
            ],
        }
        for isolate in otu["isolates"]
    ]

    return {
        "_id": otu_id,
        "name": otu["name"],
        "abbreviation": otu["abbreviation"],
        "isolates": isolates,
        "schema": otu.get("schema", []),
    }


def format_sequence_for_export(sequence: Document) -> Document:
    """Prepare a raw sequence document for export.

    If the sequence has a remote ID use this as the `_id` for export. This makes the
    sequence compatible with the original remote reference.

    :param sequence: a sequence document
    :return: the formatted sequence document
    """
    try:
        sequence_id = sequence["remote"]["id"]
    except (KeyError, TypeError):
        sequence_id = sequence["_id"]

    cleaned_sequence = {
        "_id": sequence_id,
        "accession": sequence["accession"],
        "definition": sequence["definition"],
        "host": sequence["host"],
        "sequence": sequence["sequence"],
    }

    for key in ["segment", "target"]:
        try:
            return {**cleaned_sequence, key: sequence[key]}
        except KeyError:
            pass

    return cleaned_sequence


def get_index_file_type_from_name(file_name: str) -> IndexType:
    if ".json" in file_name:
        return IndexType.json

    if ".fa" in file_name:
        return IndexType.fasta

    if ".bt" in file_name:
        return IndexType.bowtie2

    raise ValueError(f"Filename does not map to valid IndexType: {file_name}")
