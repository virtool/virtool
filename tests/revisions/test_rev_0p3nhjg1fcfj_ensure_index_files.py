"""ensure index files

Revision ID: 0p3nhjg1fcfj
Date: 2024-05-22 20:47:09.866326

"""

import gzip
from asyncio import gather
from pathlib import Path

import arrow
import pytest
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from assets.revisions.rev_0p3nhjg1fcfj_ensure_index_files import upgrade
from virtool.indexes.models import IndexType, SQLIndexFile
from virtool.migration import MigrationContext

# Revision identifiers.
name = "ensure index files"
created_at = arrow.get("2024-05-22 20:47:09.866326")
revision_id = "0p3nhjg1fcfj"

alembic_down_revision = None
virtool_down_revision = "ulapmx8i3vpg"


# Change this if an Alembic revision is required to run this migration.
required_alembic_revision = None


def get_index_file_type_from_name(file_name: str) -> IndexType:
    if ".json" in file_name:
        return IndexType.json

    if ".fa" in file_name:
        return IndexType.fasta

    if ".bt" in file_name:
        return IndexType.bowtie2

    raise ValueError(f"Filename does not map to valid IndexType: {file_name}")


class TestUpgrade:
    """Test that SQLFiles are create as is appropriate for indexes.

    - Files exist and have been added to SQL
    - Files exist but have not been added to SQL
    - Files do not exist

    Also, ensures that an index JSON file is generated if missing.
    """

    @staticmethod
    @pytest.mark.parametrize(
        "previously_upgraded",
        [
            True,
            False,
        ],
    )
    async def test_upgrade(
        ctx: MigrationContext,
        snapshot,
        previously_upgraded,
        create_task_index,
    ):
        async with ctx.pg.begin() as conn:
            await conn.run_sync(SQLIndexFile.metadata.create_all)
            await conn.commit()

        task_index = await create_task_index(ctx)

        test_dir = (
            ctx.data_path / "references" / task_index["reference"]["id"] / "index_1"
        )

        test_dir.joinpath("reference.fa.gz").write_text("FASTA file")
        test_dir.joinpath("reference.1.bt2").write_text("Bowtie2 file")

        if previously_upgraded:
            with gzip.open(test_dir / "reference.json.gz", "wt") as file:
                file.write("Complete index json")

            async with AsyncSession(ctx.pg) as session:
                session.add_all(
                    [
                        SQLIndexFile(
                            name=f"previously upgraded {path.name}",
                            index="index_1",
                            type=get_index_file_type_from_name(path.name),
                            size=5,
                        )
                        for path in sorted(test_dir.iterdir())
                    ],
                )
                await session.commit()

        await upgrade(ctx)

        async with AsyncSession(ctx.pg) as session:
            assert (
                await session.execute(select(SQLIndexFile))
            ).scalars().all() == snapshot

        with gzip.open(Path(test_dir) / "reference.json.gz", "rt") as f:
            assert f.read() == snapshot(name="json")

    @staticmethod
    async def test_upgrade_no_files(
        ctx: MigrationContext,
        snapshot,
        create_task_index,
    ):
        async with ctx.pg.begin() as conn:
            await conn.run_sync(SQLIndexFile.metadata.create_all)
            await conn.commit()

        task_index = await create_task_index(ctx)

        await ctx.mongo.indexes.update_one(
            {"_id": "index_1"},
            {"$unset": {"manifest": ""}},
        )

        test_dir = (
            ctx.data_path / "references" / task_index["reference"]["id"] / "index_1"
        )
        test_dir.rmdir()

        await upgrade(ctx)

        async with AsyncSession(ctx.pg) as session:
            assert (
                await session.execute(select(SQLIndexFile))
            ).scalars().all() == snapshot

        with (
            pytest.raises(FileNotFoundError),
            gzip.open(Path(test_dir) / "reference.json.gz", "rt") as f,
        ):
            f.read()

    @staticmethod
    async def test_upgrade_not_ready(
        ctx: MigrationContext,
        snapshot,
        create_task_index,
    ):
        async with ctx.pg.begin() as conn:
            await conn.run_sync(SQLIndexFile.metadata.create_all)
            await conn.commit()

        task_index = await create_task_index(ctx)

        await ctx.mongo.indexes.update_one(
            {"_id": "index_1"},
            {"$set": {"ready": False}},
        )

        await upgrade(ctx)

        async with AsyncSession(ctx.pg) as session:
            assert (
                await session.execute(select(SQLIndexFile))
            ).scalars().all() == snapshot

        with (
            pytest.raises(FileNotFoundError),
            gzip.open(
                Path(
                    ctx.data_path
                    / "references"
                    / task_index["reference"]["id"]
                    / "index_1",
                )
                / "reference.json.gz",
                "rt",
            ) as f,
        ):
            f.read()


@pytest.fixture()
async def create_task_index(
    config,
    reference,
    test_otu,
    test_sequence,
):
    async def func(ctx: MigrationContext):
        test_sequence["accession"] = "KX269872"
        ref_id = test_otu["reference"]["id"]

        index = {
            "_id": "index_1",
            "name": "Index 1",
            "deleted": False,
            "manifest": {test_otu["_id"]: test_otu["version"]},
            "ready": True,
            "reference": {"id": ref_id},
        }

        await gather(
            ctx.mongo.otus.insert_one(test_otu),
            ctx.mongo.sequences.insert_one(test_sequence),
            ctx.mongo.references.insert_one({**reference, "_id": ref_id}),
            ctx.mongo.indexes.insert_one(index),
        )

        index_dir = config.data_path / "references" / ref_id / index["_id"]
        index_dir.mkdir(parents=True)

        return index

    return func
