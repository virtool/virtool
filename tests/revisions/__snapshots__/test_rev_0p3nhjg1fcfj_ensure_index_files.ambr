# serializer version: 1
# name: TestUpgrade.test_upgrade[False]
  list([
    <SQLIndexFile(id=1, name=reference.1.bt2, index=index_1, type=bowtie2, size=12)>,
    <SQLIndexFile(id=2, name=reference.fa.gz, index=index_1, type=fasta, size=10)>,
    <SQLIndexFile(id=3, name=reference.json.gz, index=index_1, type=json, size=322)>,
  ])
# ---
# name: TestUpgrade.test_upgrade[False][json]
  '{"data_type":"genome","organism":"virus","otus":[{"_id":"6116cba1","name":"Prunus virus F","abbreviation":"PVF","isolates":[{"id":"cab8b360","default":true,"source_name":"8816-v2","source_type":"isolate","sequences":[{"_id":"abcd1234","accession":"KX269872","definition":"Prunus virus F isolate 8816-s2 segment RNA2 polyprotein 2 gene, complete cds.","host":"sweet cherry","sequence":"TGTTTAAGAGATTAAACAACCGCTTTC","segment":null}]}],"schema":[]}],"targets":null}'
# ---
# name: TestUpgrade.test_upgrade[True]
  list([
    <SQLIndexFile(id=1, name=previously upgraded reference.1.bt2, index=index_1, type=bowtie2, size=5)>,
    <SQLIndexFile(id=2, name=previously upgraded reference.fa.gz, index=index_1, type=fasta, size=5)>,
    <SQLIndexFile(id=3, name=previously upgraded reference.json.gz, index=index_1, type=json, size=5)>,
  ])
# ---
# name: TestUpgrade.test_upgrade[True][json]
  'Complete index json'
# ---
# name: TestUpgrade.test_upgrade_no_files
  list([
  ])
# ---
# name: TestUpgrade.test_upgrade_not_ready
  list([
  ])
# ---
